README.md
Paste Image To Markdown
=======================

A Vim/NeoVim plugin use to autosave  image from clipboard to local system storage or upload image to cloud storage which from your clipboard, and then insert link snippet to your article

## Table of Contents

  - [Screencasts](#screencasts)
  - [Features](#features)
  - [Installation](#installation)
  - [Usage](#usage)
  - [Configuration](#configuration)
  - [Examples](#examples)
  - [ChangeLog](#changelog)
  - [License](#license)

## Screencasts

**paste image from clipboard**

![paste-image-from-clipboard](https://user-images.githubusercontent.com/55470045/98808447-164f9100-2457-11eb-9550-47f176f358b3.gif)


**paste image from local path**

![paste-image-from-local-path](https://user-images.githubusercontent.com/55470045/98812660-d63fdc80-245d-11eb-9bb8-b6d48abb7353.gif)

**paste image from web url**

![paste-image-from-web-url](https://user-images.githubusercontent.com/55470045/98817049-8284c180-2464-11eb-8e2e-e7e778dcf182.gif)

**upload image from clipboard to cloud storage**

![upload-image-from-clipboard-to-cloud](https://user-images.githubusercontent.com/55470045/98820382-0d67bb00-2469-11eb-8cce-95947c3fdc39.gif)


## Features

Both of `Vim` and `NeoVim` can use this plugin

**You can upload image** from

* [x] Clipboard
* [x] Local file(path)
* [x] Remote Url

**Support OS**

* [x] MacOS
* [x] Windows
* [x] Linux
* [ ] WSL(Windows Subsystem For Linux)

**Support Markup Language**

* [x] [Markdown](https://daringfireball.net/projects/markdown/)
* [x] [Vimwiki](https://github.com/vimwiki/vimwiki)
* [x] [reStructuredText](https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html)

**Support Storage**

* [x] Local Storage
* [x] Cloud Storage (via Picgo-Core, Picgo App or uPic etc...)

**Support Cloud Storage**

> Pi2md supperts upload images to your cloud storage by 3rd-party libraries 
> like *picgo-core*, *uPic*...

* [x] PicGo-Core(nodejd moudle, recommend)
* [x] PicGo App
* [x] uPic(only on osx)


 [Picgo](https://github.com/Molunerfinn/PicGo) or [PicGo-Core](https://github.com/PicGo/PicGo-Core)

* aliyun oss
* upyun
* qiniu
* tencent cos
* Imgur
* smms
* github

[uPic](https://github.com/gee1k/uPic)

* [smms](https://sm.ms/)
* [UPYUN USS](https://www.upyun.com/products/file-storage)
* [qiniu KODO](https://www.qiniu.com/products/kodo)
* [Aliyun OSS](https://www.aliyun.com/product/oss/)
* [TencentCloud COS](https://cloud.tencent.com/product/cos)
* [BaiduCloud BOS](https://cloud.baidu.com/product/bos.html)
* [Weibo](https://weibo.com/)
* [Github](https://github.com/settings/tokens)
* [Gitee](https://gitee.com/profile/personal_access_tokens)
* [Amazon S3](https://aws.amazon.com/cn/s3/)
* [Imgur](https://imgur.com/)
* [custom upload api](https://blog.svend.cc/upic/tutorials/custom)

## Installation

**Requirements**

* vim 8.1+ and compiled with python3 support
* install `Pillow` for python3(Optional)
* cloud library(Optional)


**Install Pillow(Optional)**

You must install python3 and enable python3 for your vim/neovim
support for your vim, and then install [Pillow](https://github.com/python-pillow/Pillow) 

For `Win32/64` or `Win10` User, before `pip install`
```
python -m ensurepip
python -m pip install -U pip
python -m pip install pillow
```
Unix like platforms:
```
pip install -U pip
pip install pillow
```


**Install Picgo-Core(Optional)**

`Picgo-core` is an *nodejs* package, If you upload to cloud with picgo-core,
you can install it simply by:
```
npm install picgo -g
```


**Install Pi2md**

You can install `Pi2md` by the 3rd-party plugin managers

[vim-plug](https://github.com/junegunn/vim-plug)

```vim
call plug#begin('~/.vim/plugged')
Plug 'wancocoding/pi2md'
call plug#end()
```
after restart vim/nvim or `:so[urce]` your `vimrc`, then`:PlugInstall`

[pathogen.vim](https://github.com/tpope/vim-pathogen)

```
git clone https://github.com/wancocoding/pi2md.git ~/.vim/bundle/pi2md
```

[Vundle.vim](https://github.com/VundleVim/Vundle.vim)

```vim
call vundle#begin()
Plugin 'preservim/nerdtree'
call vundle#end()
```

[dein.vim](https://github.com/Shougo/dein.vim)

```vim
call dein#begin()
call dein#add('preservim/nerdtree')
call dein#end()
```



## Usage


You can paste image from clipboard by the command:
```
:Pi2md
```
Commands:

| Command                 | Intro                            |
| ----------------------- | -------------------------------- |
| :Pi2md                  | paste image from clipboard       |
| :Pi2md p                | paste image from a absolute path |
| :Pi2md r                | paste image from a remote url    |



### Local Storage Settings

Local Storage is the default way to save clipboard image

```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 0,
    \ 'storage_local_position_type': 0
  \ }
```

### Cloud Storage Settings

#### 1 use picgo-core(recommended)

put this in your `vimrc` file

```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 1,
    \ 'storage_cloud_tool': 'picgo-core',
    \ 'storage_cloud_picgocore_path': 
    \     '/home/vincent/.nvm/versions/node/v12.19.0/lib/node_modules/picgo/bin/picgo'
  \ }
```

config your picgo-core,see: [Picgo-core](https://github.com/PicGo/PicGo-Core)

picgo-core settings example(for aliyun oss):

edit your `~/.picgo/config.json`

```json
{
  "picBed": {
    "uploader": "aliyun",
    "current": "smms",
    "aliyun": {
      "accessKeyId": "your ak id",
      "accessKeySecret": "your ak key",
      "bucket": "your bucket name",
      "area": "your oss region",
      "path": "cocoding/blog/images/",
      "customUrl": "http://your.oss.domain.com",
      "options": ""
    }
  },
  "picgoPlugins": {
    "picgo-plugin-rename-file": true
  },
  "picgo-plugin-rename-file": {
    "format": "{y}/{m}/{d}/{hash}-{timestamp}"
  }
}
```



#### 2 use picgo app

If you use `Picgo app` as a cloud library

see: [Picgo Guide - Configuration](https://picgo.github.io/PicGo-Doc/en/guide/config.html)

```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 1,
    \ 'storage_cloud_tool': 'picgo',
    \ 'storage_cloud_picgoapp_api_server_port': '36677'
  \ }
```

Where is the picgo config file

* windows: `%APPDATA%\picgo\data.json`
* Linux: `$XDG_CONFIG_HOME/picgo/data.json` or `~/.config/picgo/data.json`
* macOS: `~/Library/Application\ Support/picgo/data.json`

An exaple on windows picgo app, `data.json`
```

{
  "uploaded": [],
  "picBed": {
    "current": "aliyun",
    "uploader": "aliyun",
    "smms": {
      "token": ""
    },
    "aliyun": {
      "accessKeyId": "your akid",
      "accessKeySecret": "your aks",
      "area": "your oss region",
      "bucket": "your bucket name",
      "customUrl": "http://yourdomain",
      "options": "",
      "path": "your/custom/path"
    }
  },
  "settings": {
    "shortKey": {
      "picgo:upload": {
        "enable": true,
        "key": "CommandOrControl+Shift+P",
        "name": "upload",
        "label": "upload-key"
      }
    },
    "server": {
      "port": 36677,
      "host": "127.0.0.1",
      "enable": true
    },
    "showUpdateTip": true,
    "customLink": "$url"
  },
  "picgoPlugins": {
    "picgo-plugin-rename-file": true
  },
  "debug": true,
  "PICGO_ENV": "GUI",
  "needReload": false,
  "picgo-plugin-rename-file": {
    "format": "{y}/{m}/{d}/{hash}-{timestamp}"
  }
}
```

then start the picgo app, make sure picgo server started with port `36677` or what your defined in your vimrc files

## Configuration

Configuring `Pi2md` is very simple, you only need to add the following content
to your `vimrc`

```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 1,
    \ 'storage_local_position_type': 0,
    \ 'storage_local_dir_name': 'images',
    \ 'storage_local_absolute_path': '/home/vincent/temp/imgs',
    \ 'storage_local_prefer_relative_path': 1,
    \ 'storage_cloud_tool': 'picgo-core',
    \ 'storage_cloud_picgocore_path': 
    \     '/home/vincent/.nvm/versions/node/v12.19.0/lib/node_modules/picgo/bin/picgo',
    \ 'storage_cloud_picgoapp_api_server_port': '36677'
  \ }
```

**debug**

- type: integer
- default: 1

Output debugging mesage if it turns on

**storage**

- type: integer
- default: 0
	+ 0: save the image locally
	+ 1: save the image in the cloud storage

**storage_local_position_type**

- type: integer
- default: 0
	+ 0: save images in `images` folder which in current directory
	+ 1: save images in a absolute path you defined

**storage_local_dir_name**

- type: string
- default: images

define the directory name for save images, when the **storage_local_position_type** value is **0**

**storage_local_absolute_path**

- type: string
- no default

> you must define a path if the **storage_local_position_type** is **1**

**storage_local_prefer_relative_path**


- type: integer
- default: 1

> Try to use relative path to display images as much as possible

**storage_cloud_tool**

- type: string
- default: picgo-core
	- picgo: a app on all platform
	- picgo-core: nodejs library
	- upic: a app on macOS

select a cloud library to save your images

**storage_cloud_picgocore_path**

- type: string
- no default

define the path of `picgo-core`, if you use `picgo-core` to save image in your cloud storage.


**storage_cloud_picgoapp_api_server_port**

- type: string
- default: 36677

picgo app start a api server, we can use it to upload image, it only works when
you use picgo app as a tool for uploading images to your cloud storage.

## Examples

### 1. Save image in current folder and use relative path

```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 0,
    \ 'storage_local_position_type': 0,
    \ 'storage_local_dir_name': 'images',
    \ 'storage_local_prefer_relative_path': 1,
  \ }
```

results:
```markdown
![Image](images/2020-10-13-22-33-02.png)
```

### 2. Save image in current folder and use absolute path

```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 0,
    \ 'storage_local_position_type': 1,
    \ 'storage_local_absolute_path': '/home/vincent/temp/imgs',
    \ 'storage_local_prefer_relative_path': 0,

  \ }
```

result:
```markdown
![Image](/Users/vincent/imgs/2020-10-13-22-37-04.png)
```

### 3. Save image in a specific directory and use relative path


```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 0,
    \ 'storage_local_position_type': 1,
    \ 'storage_local_absolute_path': 'f:\Dropbox\Dropbox\docs\md\notes\files\images\202010',
    \ 'storage_local_prefer_relative_path': 1,
  \ }
```

result:
```
![Image](..\..\..\files\images\202010\1c353706-c5a1-4865-b767-9f37e5c71c4e-2020-10-16-10-19-28.png)
```

### 4. Save image to cloud by picgo-core


```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 1,
    \ 'storage_cloud_tool': 'picgo-core',
    \ 'storage_cloud_picgocore_path': 
    \     '/home/vincent/.nvm/versions/node/v12.19.0/lib/node_modules/picgo/bin/picgo'
  \ }
```

results:
```
![Image](http://cdn.images.myclouddomain.com/cocoding/blog/images/2020/11/13/10068c58c929bcbf80598fed427dab61-1605260574.png)
```

### 5. Save image to cloud by picgo app

```vim
let g:pi2mdSettings = {
    \ 'debug': 1,
    \ 'storage': 1,
    \ 'storage_cloud_tool': 'picgo',
    \ 'storage_cloud_picgoapp_api_server_port': '36677'
  \ }
```

results:
```
![Image](http://cdn.images.myclouddomain.com/cocoding/blog/images/2020/11/13/10068c58c929bcbf80598fed427dab61-1605260574.png)
```


## ChangeLog

see: [CHANGELOG](https://github.com/wancocoding/pi2md/blob/master/CHANGELOG.md)

## License

[MIT](https://github.com/wancocoding/pi2md/blob/master/LICENSE)

Copyright © 2020 Vincent Wancocoding

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
ChangeLog
======
* 2020-11-17
	- update function of check vim version
	- update job start function of delete temp file for Neovim
* 2020-11-11 1.0.2.a1
	- rewrite readme
* 2020-11-10
	- update settings about local storage
	- bug fix
* 2020-11-06
	- fix delete temp file bug on linux, test on linux success
* 2020-11-03
	- finish insert image from remote url
* 2020-11-02
	- call init function when load plugin
	- move unnecessarily function to autoload
	- delete temp file if picgo server is not available
	- delete file repeat 3 times
	- use quickfix window to show error
	- fnameescape check
	- use job to delete temp file, because temp maybe locked by some app like dropbox
	- fix local storage not delete temp file
	- make sure file path safe on different platform
	- use python instead of nodejs script
	- remove python script inside vim script
* 2020-10-29
	- update node script
	- add function to detect picgo api service
	- set default value for not required config which has depneds and it has been setup already
	- picgo app support local image upload
	- set shell for vim script before run `system` or `systemlist` 
	- use cmd as the default shell on win32 platform
* 2020-10-28
	- add local image path support
	- Compatible with neovim function argument
* 2020-10-23
	- fix invalid argument for Neovim
	- finish upload image by picgo app with api
* 2020-10-21
	- fix win32 power shell output with special char
	- update execute command quotes
* 2020-10-20
	- refactor continue
	- text with double quotes to single quotes
* 2020-10-19
	- add check python envrionment function and error catch for python or python module
	- update check settings function
	- update settings data structure
	- refactored the functions of cloud storage
	- update readme examples
	- use try catch to handle exceptions
	- modify logging format
* 2020-10-18
	- add functions for check configuration
	- add functions for print different messages
	- refactor all functions to dict function
* 2020-10-16
	- change pi2md configuration to dictionary
	- refactor all functions
	- support reStructuredText language
	- support vimwiki markup language
	- update examples
	- update config introduction
* 2020-10-16 1.0.0.b7
	- refactor clipbord function
	- update readme with more example
* 2020-10-16 1.0.0.b6
	- add debug mode
	- use absolute path on windows when the drive are different
	- update random func, use uuid
	- use py3 script to handle clipboard on all platforms
* 2020-10-14 1.0.0.beta5
	- fix windows 10 clipboard paste bug
	- add a powershell script
	- put all script together
* 2020-10-13 1.0.0.beta4
	- add local storage for windows
	- add command
	- update readme 
* 2020-10-09 1.0.0.beta3
	- add picgo support
	- add node script for post request 
* 2020-10-09 1.0.0.beta.2
	- fix relative path bug
* 2020-10-09 1.0.0.beta
	- add support for PicGo-Core
	- add support for uPic on MacOS only
* 2020-10-08
	- you can use relative path when save to local storage


python3/pi2md/test_pi2md.py
import unittest
from pi2md import simple_hash_text, detect_file, download_image
import os
from pathlib import Path


class TestPi2md(unittest.TestCase):

    def test_simple_hash_text(self):
        hash = simple_hash_text('')
        self.assertEqual(len(hash), 32)


    def test_simple_hash_text2(self):
        to_hash = 'f:\Dropbox\Dropbox\docs\md\notes\Diary\2020\11\78e63c5a-66e1-40d6-adc4-23b4accd219b.jpg'
        hash = simple_hash_text(to_hash)
        self.assertEqual(len(hash), 32)


    def test_detect_file(self):
        fname = 'd:\\test2.txt'
        self.assertEqual(detect_file(fname), 1)
        self.assertEqual(detect_file('d:\\akhsdkasjdhjkashwyqgdqbwd'), 0)


    def test_download_image(self):
        p_path = 'f:\\downloads\\temp\\0000.jpg'
        fname = 'https://ss1.bdstatic.com/70cFuXSh_Q1YnxGkpoWK1HF6hhy/it/u=2115546205,172279996&fm=11&gp=0.jpg'
        download_image(fname, p_path)
        fp = Path(p_path)
        self.assertTrue(fp.exists())
        os.remove(p_path)


if __name__ == "__main__":
    unittest.main()


scripts/picgo.js
var axios = require('axios');

var host = '127.0.0.1'
var port = '36677'
var uploadUrl = 'http://' + host + ':' + port + '/upload'
var uploadFile = ''
var postData = ''

const args = process.argv;


if (args.indexOf('-p') > -1) {
	port = args[args.indexOf('-p') + 1]
	uploadUrl = 'http://' + host + ':' + port + '/upload'
}

if (args.indexOf('-f') > -1) {
	uploadFile = args[args.indexOf('-f') + 1]
	postData = JSON.stringify({"list": [uploadFile]})
}

function getPostConfig() {
	if (uploadFile == ''){
		var config = {
			method: 'post',
			url: uploadUrl
		};
		return config
	} else {
		var config = {
			method: 'post',
			url: uploadUrl,
			headers: {
				'Content-Type': 'application-json'
			},
			data: postData
		};
		return config
	}
}


var config = getPostConfig()

axios(config)
	.then(function (response) {
		var resultObject = response.data;
		if (resultObject['success'] == true) {
			var results = resultObject.result
			console.log(results[0])
		} else {
			console.log("error")
		}
		// console.log(JSON.stringify(response.data));
	})
	.catch(function () {
		console.log("error");
	});



scripts/windowsclipboard.ps1
# Adapted from https://github.com/octan3/img-clipboard-dump/blob/master/dump-clipboard-png.ps1
param($imagePath)

# https://github.com/PowerShell/PowerShell/issues/7233
# fix the output encoding bug
[console]::InputEncoding = [console]::OutputEncoding = New-Object System.Text.UTF8Encoding

Add-Type -Assembly PresentationCore
function main {
    $img = [Windows.Clipboard]::GetImage()

    if ($img -eq $null) {
        "no image"
        Exit 1
    }

    if (-not $imagePath) {
        "no image"
        Exit 1
    }

    $fcb = new-object Windows.Media.Imaging.FormatConvertedBitmap($img, [Windows.Media.PixelFormats]::Rgb24, $null, 0)
    $stream = [IO.File]::Open($imagePath, "OpenOrCreate")
    $encoder = New-Object Windows.Media.Imaging.PngBitmapEncoder
    $encoder.Frames.Add([Windows.Media.Imaging.BitmapFrame]::Create($fcb)) | out-null
    $encoder.Save($stream) | out-null
    $stream.Dispose() | out-null

    $imagePath
    # fix windows 10 native cmd crash bug when "picgo upload"
    # https://github.com/PicGo/PicGo-Core/issues/32
    Exit 1
}

try {
    # For WIN10
    $file = Get-Clipboard -Format FileDropList
    if ($file -ne $null) {
        Convert-Path $file
        Exit 1
    }
} catch {
    # For WIN7 WIN8 WIN10
    main
}

main


python3/pi2md/pi2md.py
#!/usr/bin/env python
# encoding: utf-8

import os
from pathlib import Path
import urllib.request
import uuid
import hashlib
import socket
import shutil
import http.client
import json
from sys import platform
from PIL import ImageGrab

IMGFOLDER = os.getcwd() + '/images/'
ip_addr = '127.0.0.1'
port = '36677'

def download_image(image_url, temp_path):
    """download a remote image url, and save a file"""
    # uuid_name = str(uuid.uuid4())
    # image_postfix = image_url.split('.')[-1]
    # temp_image_name = "".join([parent_path, os.path.sep, uuid_name, \
    #     '.', image_postfix])
    urllib.request.urlretrieve(image_url, temp_path)


def detect_picgo_api_server(api_port=port):
    """detect the picgo api server port"""
    detectSocket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    detectSocket.settimeout(3)

    detectTarget = (ip_addr, int(api_port))

    serverPortAvailable = 'NO'
    try:
        result = detectSocket.connect_ex(detectTarget)
        if result == 0:
            # connect to server success
            serverPortAvailable = 'YES'
        else:
            serverPortAvailable = 'NO'
    finally:
        detectSocket.close()
    return serverPortAvailable


def copy_file(source, dest, del_source='y'):
    if del_source == 'y':
        shutil.move(source, dest)
    else:
        shutil.copyfile(source, dest)


def remove_file(source):
    os.remove(source)


def picgo_upload(image_path=None, api_port=port):
    headers = {'Content-Type': 'application/json'}
    apiserver = 'http://' + ip_addr + ':' + api_port + '/upload'
    if image_path:
        # upload image from local image path
        post_data = {"list": [image_path]}
        params = json.dumps(post_data).encode('utf8')
        req = urllib.request \
                .Request(apiserver, data=params,
                        headers={'content-type': 'application/json'},
                        method='POST')
    else:
        # upload from system clipboard
        req = urllib.request.Request(apiserver, method='POST')
    response = urllib.request.urlopen(req)
    result = response.read().decode('utf-8')
    data = json.loads(result)
    if data['success']:
        image_url = data['result'][0]
        return image_url
    return ''


def save_clipboard(dist):
    """save the image from system clipboard"""
    tmpimg = ImageGrab.grabclipboard()
    if tmpimg:
        tmpimg.save(dist, 'PNG', compress_level=9)
        return dist
    return ''


def detect_os():
    if platform.startswith('linux'):
        return 'linux'
    if platform.startswith('freebsd'):
        return 'freebsd'
    if platform.startswith('aix'):
        return 'aix'
    if platform.startswith('win'):
        return 'windows'
    if platform.startswith('darwin'):
        return 'darwin'
    if platform.startswith('cygwin'):
        return 'cygwin'
    return ''


def detect_file(file_path):
    fp = Path(file_path)
    if fp.exists():
        return 1
    return 0


def simple_hash_text(source, hash_fn='b2s'):
    if hash_fn == 'b2s':
        ho = hashlib.blake2s(source.encode(), digest_size=16)
        return ho.hexdigest()
    if hash_fn == 'md5':
        ho = hashlib.md5(source.encode())
        return ho.hexdigest()
    if hash_fn == 'sha1':
        ho = hashlib.sha1(source.encode())
        return ho.hexdigest()
    if hash_fn == 'sha224':
        ho = hashlib.sha224(source.encode())
        return ho.hexdigest()
    if hash_fn == 'sha256':
        ho = hashlib.sha256(source.encode())
        return ho.hexdigest()
    if hash_fn == 'sha384':
        ho = hashlib.sha384(source.encode())
        return ho.hexdigest()
    if hash_fn == 'sha512':
        ho = hashlib.sha512(source.encode())
        return ho.hexdigest()
    return source


# if __name__ == '__main__':
    # download_image(
    #     'http://files.static.tiqiua.com/cocoding/blog/images/2020/10/30/08820864f35d2c405e8ae42aa7007ecb-1604026998.jpg')
    # detect_picgo_api_server('36678')
    # copy_file('d:\\develop\\workspace\\dotfiles\\windows\\vimfiles.symlink\\plugged\\pi2md\\python3\\pi2md\\f2b4c197-d1f8-4236-9d13-a12322f1d9d8.jpg',
    #     'd:\\asdasdasdasdsadasd.jpg')
    # remove_file('d:\\asdasdasdasdsadasd.jpg')
    # picgo_upload(api_port='36678')
    # save_clipboard('d:\\test.png')
    # print(simple_hash_text('hsagdfhgas.sajdpf;lsakdfpi', 'b2s'))

# vim:set ft=python fenc=utf-8 noet sts=4 sw=4 ts=4 tw=79:
